tests/functional/test_snap_dcgm.py

```diff
@@ -1,86 +1,138 @@
-
-
-@pytest.mark.parametrize(
-    "service, config, new_value",
-    [
-        ("dcgm.dcgm-exporter", "dcgm-exporter-address", ":9466"),
-        ("dcgm.nv-hostengine", "nv-hostengine-port", "5566"),
-    ],
-)
-def test_dcgm_bind_config(service: str, config: str, new_value: str):
-    """Test snap bind configuration."""
-    result = subprocess.run(
-        "sudo snap get dcgm -d".split(), check=True, capture_output=True, text=True
-    )
-    dcgm_snap_config = json.loads(result.stdout.strip())
-    assert config in dcgm_snap_config, f"{config} is not in the snap configuration"
-    old_value = dcgm_snap_config[config]
 
-    def set_config_and_check(value: str):
+class TestDCGMConfigs:
+    @classmethod
+    def set_config(cls, service: str, config: str, value: str = "") -> None:
+        """Set a configuration value for a snap service."""
         assert 0 == subprocess.call(
             f"sudo snap set dcgm {config}={value}".split()
-        ), f"Failed to set {config} to {new_value}"
+        ), f"Failed to set {config} to {value}"
 
         # restart the service to apply the configuration
         subprocess.run(f"sudo snap restart {service}".split(), check=True)
 
+    @classmethod
+    def check_bind_config(cls, service: str, bind: str) -> None:
+        """Check if a service is listening on a specific bind."""
         for attempt in Retrying(wait=wait_fixed(2), stop=stop_after_delay(10)):
             with attempt:
                 assert 0 == subprocess.call(
-                    f"nc -z localhost {value.lstrip(':')}".split()
-                ), f"{service} is not listening on {value}"
+                    f"nc -z localhost {bind.lstrip(':')}".split()
+                ), f"{service} is not listening on {bind}"
+
+    @classmethod
+    def check_config_exists(cls, config: str) -> str:
+        """Check if a configuration exists in the snap configuration.
+
+        :return: The value of the current configuration
+        """
+        result = subprocess.check_output("sudo snap get dcgm -d".split(), text=True)
+        dcgm_snap_config = json.loads(result.strip())
+        assert config in dcgm_snap_config, f"{config} is not in the snap configuration"
+        return str(dcgm_snap_config[config])
+
+    @classmethod
+    @retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+    def check_metric_config(cls, metric_file: str) -> None:
+        dcgm_exporter_service = "snap.dcgm.dcgm-exporter"
+
+        result = subprocess.check_output(
+            f"sudo systemctl show -p ActiveEnterTimestamp {dcgm_exporter_service}".split(),
+            text=True,
+        )
+
+        start_timestamp = result.strip().split("=")[1]
+
+        result = subprocess.check_output(
+            f"sudo journalctl -u {dcgm_exporter_service} --since '{start_timestamp}'",
+            shell=True,
+            text=True,
+        )
+
+        assert metric_file in result, f"Metric file {metric_file} is not loaded"
```

Comment:
I investigated and instead of checking the logs you can run something like that it's a much better approach:

```shell
ps -eo cmd | grep "bin/dcgm-exporter"                                                                                                                                     ok  18:11:14 
/snap/dcgm/x1/bin/dcgm-exporter -a :9400
```
ps will show what commands were called to start the process :wink:

Reply:
Yes, but I am checking that the `dcgm-exporter` acknowledged the file, you can supply a file and it might fallback if file does exist or there is something wrong with the file itself

Reply:
I investigated a little bit and surprisingly it accepts an empty file as you are doing on the tests. I don't think that makes sense, but that is upstream choice. In our case I'm against because if you don't want to generate metrics you can simply stop the daemon service.

In any case, I think we should focus on the snap side and not much on the payload because that is upstream concern.

For those tests IMO would be enough:

- without setting a custom file the ps shows that `-f` was not used
- pass a custom valid file (not empty), restart the service to take effect, check that was callled on ps with `-f` and  prove that returns 200
- pass an invalid file (string unknown)  , restart the service to take effect, check that was callled on ps with `-f` and  prove that does not return 200

Reply:
But it will return 200 in both cases, because it will fallback to the default metrics file in `/etc/dcgm-exporter`, that is why I am testing the logs, as I want to make sure the the exporter consumed the provided file

Reply:
> But it will return 200 in both cases, because it will fallback to the default metrics file in /etc/dcgm-exporter


That is not what happens in my machine, is this happening on a VM?:

```
sudo snap logs dcgm.dcgm-exporter                                                                                    ok  18:55:32 
2024-09-17T18:55:33-03:00 dcgm.dcgm-exporter[1631362]: time="2024-09-17T18:55:33-03:00" level=info msg="DCGM successfully initialized!"
2024-09-17T18:55:33-03:00 dcgm.dcgm-exporter[1631362]: time="2024-09-17T18:55:33-03:00" level=info msg="Not collecting DCP metrics: This request is serviced by a module of DCGM that is not currently loaded"
2024-09-17T18:55:33-03:00 dcgm.dcgm-exporter[1631362]: time="2024-09-17T18:55:33-03:00" level=info msg="Falling back to metric file '/var/snap/dcgm/common/test-metrics.csv'"
2024-09-17T18:55:33-03:00 dcgm.dcgm-exporter[1631362]: time="2024-09-17T18:55:33-03:00" level=fatal msg="malformed CSV record; err: failed to parse line 0 (`[potato]`), expected 3 fields"
2024-09-17T18:55:33-03:00 systemd[1]: snap.dcgm.dcgm-exporter.service: Main process exited, code=exited, status=1/FAILURE
2024-09-17T18:55:33-03:00 systemd[1]: snap.dcgm.dcgm-exporter.service: Failed with result 'exit-code'.
2024-09-17T18:55:33-03:00 systemd[1]: snap.dcgm.dcgm-exporter.service: Scheduled restart job, restart counter is at 5.
2024-09-17T18:55:33-03:00 systemd[1]: snap.dcgm.dcgm-exporter.service: Start request repeated too quickly.
2024-09-17T18:55:33-03:00 systemd[1]: snap.dcgm.dcgm-exporter.service: Failed with result 'exit-code'.
2024-09-17T18:55:33-03:00 systemd[1]: Failed to start snap.dcgm.dcgm-exporter.service - Service for snap application dcgm.dcgm-exporter.
 !w /var/snap/dcgm/common  curl localhost:9400/metrics                                                                                          ok  18:55:36 
curl: (7) Failed to connect to localhost port 9400 after 0 ms: Couldn't connect to server

```

Reply:
I see what you mean

Reply:
adding to the thread, I think Gabriel is right that as packagers, our scope ends at "did I pass the payload the correct arguments?". So just checking the cmdline would be sufficient

Reply:
also fyi another way to get to that invocation would be
```
cat /proc/$(pgrep dgcm-exporter)/cmdline
```

either works

Reply:
or even simpler: `ps -C dcgm-exporter -o cmd`